jocd/src/br/org/certi/jocd/dapaccess/connectioninterface/AndroidUsbDevice.java

```diff
@@ -295,26 +299,36 @@ public void open() {
 
     // We should never get here... unless something went wrong.
     // Clean whatever we done here.
-    this.deviceConnection = null;
-    this.inputEndpoint = null;
-    this.outputEndpoint = null;
-    this.interfaceNumber = -1;
-    this.open = false;
+    close();
   }
 
   /*
    * Close the device.
    */
   public void close() {
-    if (connection == null) {
-      return;
-    }
+    // Stop the rx thread (if exists).
     if (rxThread != null) {
       stopRxThread();
     }
-    connection.releaseInterface(usbInterface);
-    open = false;
-    connection.close();
+
+    // Release and close devise connection.
+    if (this.deviceConnection != null) {
+      this.deviceConnection.releaseInterface(usbInterface);
+      this.deviceConnection.close();
+      this.deviceConnection = null;
+    }
+
+    // Clean endpoints and interface number.
+    this.inputEndpoint = null;
+    this.outputEndpoint = null;
+    this.interfaceNumber = -1;
+
+    // Allow this device to be opened again.
+    this.atomicOpen.set(false);
```

Comment:
Shouldn't atomicOpen be checked in the beginning of this function to avoid execution on an already closed connection?

Reply:
I think this is not necessary because there is no problem to "close" twice - all objects are being checked if is already null, and than set this to false (and at this point, doesn't matter if it was open or not... now it is closed).

In fact, if something goes wrong on "open()" we will call "close()" and clean everything.

Reply:
@ghn-certi I'll merge this pull request, and if necessary we add a issue for this later.